src/ipvs/ip_vs_service_match_acl.c

```diff
@@ -0,0 +1,929 @@
+        }
+    }
+    return 0;
+}
+
+static void *dpvs_svc_match_build_loop(void *data)
+{
+    uint64_t pre4[DPVS_MAX_LCORE] = { 0 };
+    uint64_t pre6[DPVS_MAX_LCORE] = { 0 };
+    struct timeval t1;
+    struct timeval t2;
+    struct timeval d;
+    int cid = 0;
+    int build = 0;
+    dpvs_svc_match_build_thread_init(data);
+    while (!stop_build) {
+        gettimeofday(&t1, NULL);
+        for (cid = 0; cid < DPVS_MAX_LCORE; cid++) {
+            if (pre4[cid] != dpvs_match4_generation[cid]) {
+                pre4[cid] = dpvs_match4_generation[cid];
+                build_acl(AF_INET, cid);
+                build = 1;
+            }
+            if (pre6[cid] != dpvs_match6_generation[cid]) {
+                pre6[cid] = dpvs_match6_generation[cid];
+                build_acl(AF_INET6, cid);
+                build = 1;
+            }
+        }
+        if (build) {
+            gettimeofday(&t2, NULL);
+            if (t2.tv_usec < t1.tv_usec) {
+                d.tv_usec = t2.tv_usec + 1000000 - t1.tv_usec;
+                d.tv_sec = t2.tv_sec - t1.tv_sec - 1;
+            } else {
+                d.tv_usec = t2.tv_usec - t1.tv_usec;
+                d.tv_sec = t2.tv_sec - t1.tv_sec;
+            }
+            RTE_LOG(DEBUG, IPVS, "build acl time used:%lu.%06lu,"
+                    " start at %lu.%06lu end at %lu.%06lu\n",
+                    d.tv_sec, d.tv_usec,
+                    t1.tv_sec, t1.tv_usec, t2.tv_sec, t2.tv_usec);
+        } else {
+            usleep(100000);
+        }
+    }
+    return NULL;
+}
+
+int dp_vs_svc_match_init(void)
+{
+    uint64_t i = 0;
+    int size = 0;
+    int cid = 0;
+    char name[RTE_RING_NAMESIZE];
+    for (cid = 0; cid < DPVS_MAX_LCORE; cid++) {
+        snprintf(name, RTE_RING_NAMESIZE, "dpvs_idx_q%d", cid);
+        dpvs_svc_idxq[cid] = rte_ring_create(name, rte_align32pow2(DPVS_SVC_MAX),
+                SOCKET_ID_ANY, 0);
+        if (!dpvs_svc_idxq[cid]) {
+            RTE_LOG(ERR, IPVS, "fail to create %s\n", name);
+            dp_vs_svc_match_term();
+            return EDPVS_NOMEM;
+        }
+        for (i = 1; i < DPVS_SVC_MAX; i++) {
+            rte_ring_enqueue(dpvs_svc_idxq[cid], (void*)i);
+        }
+        size = sizeof(struct dp_vs_service*) * DPVS_SVC_MAX * DPVS_MAX_LCORE;
+        dpvs_svcs[cid] = rte_malloc(NULL, size, RTE_CACHE_LINE_SIZE);
+        if (!dpvs_svcs[cid]) {
+            RTE_LOG(ERR, IPVS, "fail to alloc dpvs_svcs\n");
+            dp_vs_svc_match_term();
+            return EDPVS_NOMEM;
+        }
+        memset(dpvs_svcs[cid], 0, size);
+        dp_vs_svc_match_acl4[cid] = NULL;
+        dp_vs_svc_match_acl6[cid] = NULL;
+    }
+    stop_build = 0;
+    pthread_t tid;
+    if (pthread_create(&tid, NULL, dpvs_svc_match_build_loop, NULL)) {
+        RTE_LOG(ERR, IPVS, "fail to create dpvs_svc_match_build_loop thread\n");
+        dp_vs_svc_match_term();
+        return EDPVS_SYSCALL;
+    }
```

Comment:
DPVS is based on a poll mode driver, with each thread pinned to an exclusive cpu, and extra threading is not encouraged for performance considerations.

Reply:
10w dpdk ACL rules compiling will take about 10 seconds, this should not do in the master. a new thread is very needed.